rectools/models/candidate_ranking.py

```diff
@@ -0,0 +1,637 @@
+                "group_id": group_ids,
+            }
+        else:
+            raise ValueError("Got unexpected model_type")
+
+        if self.pool_kwargs is not None:
+            pool_kwargs.update(self.pool_kwargs)
+
+        fit_kwargs = {"X": Pool(**pool_kwargs)}
+
+        if self.fit_kwargs is not None:
+            fit_kwargs.update(self.fit_kwargs)
+
+        return fit_kwargs
+
+
+class CandidateFeatureCollector:
+    """
+    Base class for collecting features for candidates user-item pairs. Useful for creating train with features for
+    CandidateRankingModel.
+    Using this in CandidateRankingModel will result in not adding any features at all.
+    Inherit from this class and rewrite private methods to grab features from dataset and external sources
+    """
+
+    # TODO: this class can be used in pipelines directly. it will keep scores and ranks and add nothing
+    # TODO: create an inherited class that will get all features from dataset?
+
+    def _get_user_features(
+        self, users: ExternalIds, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        return pd.DataFrame(columns=[Columns.User])
+
+    def _get_item_features(
+        self, items: ExternalIds, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        return pd.DataFrame(columns=[Columns.Item])
+
+    def _get_user_item_features(
+        self, useritem: pd.DataFrame, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        return pd.DataFrame(columns=Columns.UserItem)
+
+    def collect_features(
+        self, useritem: pd.DataFrame, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        """
+        Collect features for users-item pairs from any desired sources.
+
+        Parameters
+        ----------
+        useritem : pd.DataFrame
+            Candidates with score/rank features from first stage. Ids are either external or 1x internal
+        dataset : Dataset
+            Dataset will have either external -> 2x internal id maps to internal -> 2x internal
+        fold_info : tp.Optional[tp.Dict[str, tp.Any]]
+            Fold inofo from splitter can be used for adding time-based features
+
+        Returns
+        -------
+        pd.DataFrame
+            `useritem` dataframe enriched with features for users, items and useritem pairs
+        """
+
+        user_features = self._get_user_features(useritem[Columns.User].unique(), dataset, fold_info)
+        item_features = self._get_item_features(useritem[Columns.Item].unique(), dataset, fold_info)
+        useritem_features = self._get_user_item_features(useritem, dataset, fold_info)
+
+        res = useritem
+        res = res.merge(user_features, on=Columns.User, how="left")
+        res = res.merge(item_features, on=Columns.Item, how="left")
+        res = res.merge(useritem_features, on=Columns.UserItem, how="left")
```

Comment:
It's better to use a chain of operations, otherwise to this moment you will have 4 datasets instead of 1, all are in your memory

Reply:
Fixed